contracts/core/TokenAllocator.sol

```diff
@@ -0,0 +1,269 @@
+
+        allocationInfo = AllocationInfoView(
+            id,
+            allocation.isClosed,
+            allocation.allocator,
+            allocation.token,
+            allocation.balance,
+            allocation.merkleRoot,
+            allocation.descriptionURL
+        );
+    }
+
+    function isClaimed(uint256 id, address user) external view withCorrectId(id) returns (bool) {
+        AllocationData storage allocationInfo = _allocationInfos[id];
+
+        return allocationInfo.claimed.contains(user);
+    }
+
+    function _createAllocation(
+        address allocator,
+        address token,
+        uint256 amount,
+        bytes32 merkleRoot,
+        string calldata descriptionURL
+    ) internal {
+        lastAllocationId++;
+        uint256 id = lastAllocationId;
+
+        require(token != address(0), "TA: Zero token address");
+        require(amount > 0, "TA: Zero ammount to allocate");
+        require(merkleRoot != bytes32(0), "TA: Zero Merkle root");
+
+        AllocationData storage allocationInfo = _allocationInfos[id];
+
+        allocationInfo.token = token;
+        allocationInfo.balance = amount;
+        allocationInfo.allocator = allocator;
+        allocationInfo.merkleRoot = merkleRoot;
+        allocationInfo.descriptionURL = descriptionURL;
+
+        _updateGlobalInfo(allocator, token, id);
+
+        emit AllocationCreated(id, allocator, token, amount, merkleRoot, descriptionURL);
+    }
+
+    function _updateGlobalInfo(address allocator, address token, uint256 id) internal {
+        _tokensByAllocator[allocator].add(token);
+        _allocatorsByToken[token].add(allocator);
+        _allocations[allocator][token].add(id);
+    }
+
+    function _idsToAllocationInfos(
+        uint256[] memory ids
+    ) internal view returns (AllocationInfoView[] memory infos) {
+        uint256 length = ids.length;
+        infos = new AllocationInfoView[](length);
+
+        for (uint256 i = 0; i < length; i++) {
+            infos[i] = getAllocationInfo(ids[i]);
+        }
+    }
+
+    function _arraysCopy(
+        AllocationInfoView[] memory from,
+        AllocationInfoView[] memory to,
+        uint256 startingIndex
+    ) internal pure {
+        for (uint256 i = 0; i < from.length; i++) {
+            to[startingIndex + i] = from[i];
+        }
+    }
+
+    function _retrieveAllocationData(
+        IPoolFactory.GovPoolDeployParams calldata parameters
+    ) internal view returns (address allocator, address token, uint256 amount) {
+        IPoolFactory.GovPoolPredictedAddresses memory predictedAddresses = _poolFactory
+            .predictGovAddresses(tx.origin, parameters.name);
+        allocator = predictedAddresses.govPool;
+
+        token = parameters.userKeeperParams.tokenAddress;
+        amount = 0;
+
+        IERC20Gov.ConstructorParams calldata tokenParams = parameters.tokenParams;
+        address[] calldata users = tokenParams.users;
+
+        for (uint i = 0; i < users.length; i++) {
```

Comment:
Should this only work when the govToken is the one we deploy? Then a `require(predictedAddresses.govToken == parameters.userKeeperParams.tokenAddress`, "TA: invalid token allocation address")` is missing because these parameters will just be skipped.